File: src/types/openrouter.ts
export interface OpenRouterAccount {
  id: string;
  email: string;
  /** Display-only truncated API key label from /key response, or the last 4 chars */
  label?: string | null;
  /** Key type determined from /key response or user selection */
  key_type: 'api' | 'management' | 'provisioning';
  is_free_tier: boolean;
  /** Total USD spent */
  usage: number | null;
  /** Daily usage in USD */
  usage_daily: number | null;
  /** Weekly usage in USD */
  usage_weekly: number | null;
  /** Monthly usage in USD */
  usage_monthly: number | null;
  /** Credit limit in USD */
  limit: number | null;
  /** Remaining limit in USD */
  limit_remaining: number | null;
  /** Total purchased credits (management key only) */
  total_credits: number | null;
  /** Total usage from /credits (management key only) */
  total_usage: number | null;
  /** Rate limit requests per interval */
  rate_limit_requests: number | null;
  /** Rate limit interval (e.g., "1h") */
  rate_limit_interval: string | null;
  /** Label from /key response */
  key_label: string | null;
  /** Status of the key */
  status: string | null;
  /** Reason for status */
  status_reason: string | null;
  created_at: number;
  last_used: number;
  tags?: string[] | null;
  usage_updated_at?: number | null;
  quota_query_last_error?: string | null;
  quota_query_last_error_at?: number | null;
  /** Raw /key response for debugging */
  auth_key_raw?: unknown;
  /** Raw /credits response (management key only) */
  auth_credits_raw?: unknown;
}

export interface OpenRouterModel {
  id: string;
  name: string;
  pricing: {
    prompt: string;
    completion: string;
    image?: string;
    audio?: string;
    web_search?: string;
  };
  context_length: number;
  top_provider: string;
  /** Whether this is a free model (pricing is all "0") */
  is_free: boolean;
  supported_parameters: string[];
}

export interface OpenRouterUsage {
  used: number | null;
  limit: number | null;
  remaining: number | null;
  percentage: number | null;
  daily: number | null;
  weekly: number | null;
  monthly: number | null;
}

export interface OpenRouterCreditsInfo {
  total_credits: number | null;
  total_usage: number | null;
  total_paid: number | null;
}

/** Get display email for an OpenRouter account - uses label or first 8 chars of ID */
export function getOpenRouterAccountDisplayEmail(account: OpenRouterAccount): string {
  return account.label
    || account.key_label
    || account.email
    || account.id.slice(0, 8);
}

/** Get plan badge text: "FREE", "PAID", or the key type label */
export function getOpenRouterPlanBadge(account: OpenRouterAccount): string {
  if (account.is_free_tier) return 'FREE';
  if (account.key_type === 'management') return 'MANAGEMENT';
  if (account.key_type === 'provisioning') return 'PROVISIONING';
  return 'PAID';
}

export function getOpenRouterPlanBadgeClass(account: OpenRouterAccount): string {
  if (account.is_free_tier) return 'badge-free';
  if (account.key_type === 'management') return 'badge-management';
  if (account.key_type === 'provisioning') return 'badge-provisioning';
  return 'badge-paid';
}

export function isOpenRouterFreeTier(account: OpenRouterAccount): boolean {
  return account.is_free_tier;
}

export function isOpenRouterManagementKey(account: OpenRouterAccount): boolean {
  return account.key_type === 'management';
}

/** Format USD cents for display */
export function formatOpenRouterCredits(cents: number | null): string {
  if (cents == null) return '—';
  return `$${cents.toFixed(2)}`;
}

/** Calculate usage as percentage of limit */
export function getOpenRouterUsagePercent(account: OpenRouterAccount): number | null {
  if (account.usage == null || account.limit == null || account.limit <= 0) return null;
  const pct = (account.usage / account.limit) * 100;
  return Math.max(0, Math.min(100, pct));
}

/** Get all usage info for the account - compatible with ProviderUsage */
export function getOpenRouterUsage(account: OpenRouterAccount): {
  inlineSuggestionsUsedPercent: number | null;
  chatMessagesUsedPercent: number | null;
  creditsUsed: number | null;
  creditsTotal: number | null;
  creditsRemaining: number | null;
} {
  const percentage = getOpenRouterUsagePercent(account);
  return {
    inlineSuggestionsUsedPercent: percentage != null ? Math.round(percentage) : null,
    chatMessagesUsedPercent: percentage != null ? Math.round(percentage) : null,
    creditsUsed: account.usage,
    creditsTotal: account.limit,
    creditsRemaining: account.limit_remaining,
  };
}

/** Get credits info (management key only) */
export function getOpenRouterCreditsInfo(
  account: OpenRouterAccount,
): OpenRouterCreditsInfo | null {
  if (account.key_type !== 'management') return null;
  return {
    total_credits: account.total_credits,
    total_usage: account.total_usage,
    total_paid:
      account.total_credits != null && account.total_usage != null
        ? account.total_credits - account.total_usage
        : null,
  };
}

File: src-tauri/src/models/openrouter.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterAccount {
    pub id: String,
    pub email: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(rename = "key_type")]
    pub key_type: OpenRouterKeyType,
    #[serde(rename = "is_free_tier")]
    pub is_free_tier: bool,
    #[serde(default)]
    pub usage: Option<f64>,
    #[serde(default)]
    pub usage_daily: Option<f64>,
    #[serde(default)]
    pub usage_weekly: Option<f64>,
    #[serde(default)]
    pub usage_monthly: Option<f64>,
    #[serde(default)]
    pub limit: Option<f64>,
    #[serde(default)]
    pub limit_remaining: Option<f64>,
    #[serde(default)]
    pub total_credits: Option<f64>,
    #[serde(default)]
    pub total_usage: Option<f64>,
    #[serde(default)]
    pub rate_limit_requests: Option<i64>,
    #[serde(default)]
    pub rate_limit_interval: Option<String>,
    #[serde(default)]
    pub key_label: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub status_reason: Option<String>,
    #[serde(default)]
    pub usage_updated_at: Option<i64>,
    #[serde(default)]
    pub tags: Option<Vec<String>>,
    #[serde(default)]
    pub quota_query_last_error: Option<String>,
    #[serde(default)]
    pub quota_query_last_error_at: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_key_raw: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_credits_raw: Option<Value>,
    pub created_at: i64,
    pub last_used: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum OpenRouterKeyType {
    #[serde(rename = "api")]
    Api,
    #[serde(rename = "management")]
    Management,
    #[serde(rename = "provisioning")]
    Provisioning,
}

impl OpenRouterKeyType {
    pub fn as_str(&self) -> &'static str {
        match self {
            OpenRouterKeyType::Api => "api",
            OpenRouterKeyType::Management => "management",
            OpenRouterKeyType::Provisioning => "provisioning",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "management" => OpenRouterKeyType::Management,
            "provisioning" => OpenRouterKeyType::Provisioning,
            _ => OpenRouterKeyType::Api,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterUsage {
    pub used: Option<f64>,
    pub limit: Option<f64>,
    pub remaining: Option<f64>,
    pub percentage: Option<f64>,
    pub daily: Option<f64>,
    pub weekly: Option<f64>,
    pub monthly: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterCreditInfo {
    pub total_credits: Option<f64>,
    pub total_usage: Option<f64>,
    pub total_paid: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterModel {
    pub id: String,
    pub name: String,
    pub pricing: OpenRouterModelPricing,
    pub context_length: i64,
    pub top_provider: String,
    pub is_free: bool,
    pub supported_parameters: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterModelPricing {
    pub prompt: String,
    pub completion: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub web_search: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterAccountSummary {
    pub id: String,
    pub email: String,
    pub key_type: OpenRouterKeyType,
    pub is_free_tier: bool,
    pub usage: Option<f64>,
    pub tags: Option<Vec<String>>,
    pub created_at: i64,
    pub last_used: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenRouterAccountIndex {
    pub version: String,
    pub accounts: Vec<OpenRouterAccountSummary>,
}

impl OpenRouterAccountIndex {
    pub fn new() -> Self {
        Self {
            version: "1.0".to_string(),
            accounts: Vec::new(),
        }
    }
}

impl Default for OpenRouterAccountIndex {
    fn default() -> Self {
        Self::new()
    }
}

impl OpenRouterAccount {
    pub fn summary(&self) -> OpenRouterAccountSummary {
        OpenRouterAccountSummary {
            id: self.id.clone(),
            email: self.email.clone(),
            key_type: self.key_type.clone(),
            is_free_tier: self.is_free_tier,
            usage: self.usage,
            tags: self.tags.clone(),
            created_at: self.created_at,
            last_used: self.last_used,
        }
    }
}
